src/vs/workbench/contrib/inlineChat/browser/inlineChatDecorations.ts

```diff
@@ -0,0 +1,199 @@
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Codicon } from 'vs/base/common/codicons';
+import { ThemeIcon } from 'vs/base/common/themables';
+import { ICodeEditor, IEditorMouseEvent } from 'vs/editor/browser/editorBrowser';
+import { IEditorContribution } from 'vs/editor/common/editorCommon';
+import { GlyphMarginLane, IModelDecorationsChangeAccessor, TrackedRangeStickiness } from 'vs/editor/common/model';
+import { ModelDecorationOptions } from 'vs/editor/common/model/textModel';
+import { localize } from 'vs/nls';
+import { registerIcon } from 'vs/platform/theme/common/iconRegistry';
+import { Selection } from 'vs/editor/common/core/selection';
+import { InlineChatController } from 'vs/workbench/contrib/inlineChat/browser/inlineChatController';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { DisposableStore, IDisposable } from 'vs/base/common/lifecycle';
+import { GutterActionsRegistry } from 'vs/workbench/contrib/codeEditor/browser/editorLineNumberMenu';
+import { Action } from 'vs/base/common/actions';
+import { CTX_INLINE_CHAT_VISIBLE, IInlineChatService } from 'vs/workbench/contrib/inlineChat/common/inlineChat';
+import { IContextKeyService } from 'vs/platform/contextkey/common/contextkey';
+import { ResourceMap } from 'vs/base/common/map';
+import { URI } from 'vs/base/common/uri';
+
+const gutterInlineChatIcon = registerIcon('inline-chat', Codicon.sparkle, localize('startInlineChatIcon', 'Icon which spawns the inline chat from the gutter'));
+
+export class InlineChatDecorationsContribution implements IEditorContribution {
+
+	private disposableStore = new DisposableStore();
+	private onProvidersChange: IDisposable;
+	private onConfigurationChange: IDisposable | undefined;
+	private numberOfProviders: number;
+
+	private gutterDecorationsMap = new ResourceMap<{ id: string; lineNumber: number } | undefined>();
+	private inlineChatLineNumber: number | undefined;
+
+	public static readonly gutterSettingID = 'inlineChat.showGutterIcon';
+	private static readonly gutterIconClassName = 'codicon-inline-chat';
+
+	private static readonly GUTTER_DECORATION = ModelDecorationOptions.register({
+		description: 'inline-chat-decoration',
+		glyphMarginClassName: ThemeIcon.asClassName(gutterInlineChatIcon),
+		glyphMargin: { position: GlyphMarginLane.Left },
+		stickiness: TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges,
+	});
+
+	constructor(
+		private readonly editor: ICodeEditor,
+		@IInlineChatService inlineChatService: IInlineChatService,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IContextKeyService private readonly contextKeyService: IContextKeyService
+	) {
+		this.onProvidersChange = inlineChatService.onDidChangeProviders(() => {
```

Comment:
To find out if there a provider can be done with `Iterable.isEmpty(inlineChatService.getAllProvider())`